ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/JmsXAResourceRecoveryHelper.java

```diff
@@ -177,4 +268,11 @@ private void disconnect() throws XAException {
         }
     }
 
+    private void assertIfRecoveryScanWasStarted() throws XAException {
+        if (delegate == null) {
+            LOGGER.warn("XAResource.recover has to be called first with XAResource.TMSTARTRSCAN argument");
+            throw new XAException(XAException.XAER_PROTO);
```

Comment:
potentially as we are under the control of Arjuna recovery manager we could completely ignore the check and report an NPE back as connect would be equal to null.

I think so long as the Javadoc expresses it correctly, going with PROTO and explaining its expectations will be enough. INVAL is defined as "invalid arguments were given" and in particular to xa_recover has some defined semantics. I think PROTO is closest.

Options:
1. A new special exception
2. PROTO or INVAL (or a mix)
3. No protection (as we know we are running under Narayana)

I think we should move the discussion to the forum such that people can contribute. But IMO we should go with proto or no protection.